File: api/livescores.js
// Fetches Premier League standings + recent/upcoming matches from
// football-data.org, using an API key stored as a Vercel environment
// variable (never exposed to the browser). This replaces ESPN's
// undocumented API, which frequently blocks requests coming from
// server/datacenter IPs like Vercel's.
//
// Requires an environment variable FOOTBALL_DATA_API_KEY to be set in
// your Vercel project (Settings -> Environment Variables). Get a free
// key at https://www.football-data.org/client/register

export default async function handler(req, res) {
  const apiKey = process.env.FOOTBALL_DATA_API_KEY;

  if (!apiKey) {
    res.status(500).json({ error: 'FOOTBALL_DATA_API_KEY is not configured on the server' });
    return;
  }

  const headers = { 'X-Auth-Token': apiKey };

  // Recent + upcoming matches: a small window around today so we're not
  // pulling the whole season.
  const today = new Date();
  const dateFrom = new Date(today);
  dateFrom.setDate(dateFrom.getDate() - 3);
  const dateTo = new Date(today);
  dateTo.setDate(dateTo.getDate() + 3);
  const fmt = (d) => d.toISOString().slice(0, 10);

  try {
    const [matchesRes, standingsRes] = await Promise.all([
      fetch(`https://api.football-data.org/v4/competitions/PL/matches?dateFrom=${fmt(dateFrom)}&dateTo=${fmt(dateTo)}`, { headers }),
      fetch('https://api.football-data.org/v4/competitions/PL/standings', { headers }),
    ]);

    if (!matchesRes.ok || !standingsRes.ok) {
      const status = !matchesRes.ok ? matchesRes.status : standingsRes.status;
      res.status(status).json({ error: 'football-data.org returned ' + status + ' (check your API key)' });
      return;
    }

    const matchesData = await matchesRes.json();
    const standingsData = await standingsRes.json();

    // Reshape into the ESPN-like shape the frontend's renderLiveBody()
    // already knows how to render, so no frontend changes are needed
    // beyond swapping which endpoint we call.
    const events = (matchesData.matches || []).map(m => ({
      competitions: [{
        competitors: [
          { team: { shortDisplayName: m.homeTeam.shortName || m.homeTeam.name, logo: m.homeTeam.crest }, score: m.score?.fullTime?.home ?? '-' },
          { team: { shortDisplayName: m.awayTeam.shortName || m.awayTeam.name, logo: m.awayTeam.crest }, score: m.score?.fullTime?.away ?? '-' },
        ],
        status: { type: { description: m.status } },
      }],
    }));

    const table = (standingsData.standings?.[0]?.table || []);
    const entries = table.map(row => ({
      team: { shortDisplayName: row.team.shortName || row.team.name, logos: [{ href: row.team.crest }] },
      stats: [
        { name: 'gamesPlayed', value: row.playedGames },
        { name: 'wins', value: row.won },
        { name: 'ties', value: row.draw },
        { name: 'losses', value: row.lost },
        { name: 'pointDifferential', value: row.goalDifference },
        { name: 'points', value: row.points },
      ],
    }));

    res.setHeader('Cache-Control', 's-maxage=60, stale-while-revalidate=300');
    res.status(200).json({
      scoreboard: { events },
      standings: { standings: [{ entries }] },
    });
  } catch (e) {
    res.status(502).json({ error: 'Fetch failed: ' + (e && e.message ? e.message : String(e)) });
  }
}

File: api/livescores.py
import json
import os
import time
import urllib.request
import urllib.error
import urllib.parse
from http.server import BaseHTTPRequestHandler

# Fetches Premier League standings + matches for a given gameweek (matchday)
# from football-data.org, using an API key stored as a Vercel environment
# variable (never exposed to the browser).
#
# Requires an environment variable FOOTBALL_DATA_API_KEY to be set in
# your Vercel project (Settings -> Environment Variables). Get a free
# key at https://www.football-data.org/client/register
#
# Query params:
#   ?matchday=N   optional. Defaults to the Premier League's current matchday.
#
# football-data.org's free tier has a very low per-minute rate limit shared
# across ALL visitors to this app. This is the homepage widget, so it's the
# most-hit endpoint — we retry once on a 429 and cache aggressively so a
# burst of visits only needs one upstream round trip.

# The Premier League is a fixed 20-club league, so it always plays exactly
# 38 matchdays a season. Used to clamp/bound the gameweek prev/next arrows
# on the frontend without an extra API call.
TOTAL_MATCHDAYS = 38

# Maps football-data.org's raw match status to an ESPN-like {state, description}
# pair, since the frontend's "live" pulse indicator checks state === 'in'.
STATUS_MAP = {
    'SCHEDULED': ('pre', 'Scheduled'),
    'TIMED': ('pre', 'Scheduled'),
    'IN_PLAY': ('in', 'Live'),
    'PAUSED': ('in', 'Half Time'),
    'FINISHED': ('post', 'Full Time'),
    'POSTPONED': ('post', 'Postponed'),
    'SUSPENDED': ('post', 'Suspended'),
    'CANCELLED': ('post', 'Cancelled'),
}


def _fetch_json(url, headers, retry_on_429=True):
    req = urllib.request.Request(url, headers=headers)
    try:
        with urllib.request.urlopen(req, timeout=10) as resp:
            return json.loads(resp.read().decode('utf-8'))
    except urllib.error.HTTPError as e:
        if e.code == 429 and retry_on_429:
            time.sleep(1.5)
            return _fetch_json(url, headers, retry_on_429=False)
        raise


class handler(BaseHTTPRequestHandler):
    def do_GET(self):
        api_key = os.environ.get('FOOTBALL_DATA_API_KEY')
        if not api_key:
            self._send_json(500, {'error': 'FOOTBALL_DATA_API_KEY is not configured on the server'})
            return

        headers = {'X-Auth-Token': api_key}
        query = urllib.parse.urlparse(self.path).query
        params = urllib.parse.parse_qs(query)
        requested_matchday = (params.get('matchday') or [None])[0]

        try:
            matchday = int(requested_matchday) if requested_matchday else None
        except ValueError:
            matchday = None

        try:
            if matchday is None:
                comp = _fetch_json('https://api.football-data.org/v4/competitions/PL', headers)
                matchday = comp.get('currentSeason', {}).get('currentMatchday') or 1
            matchday = max(1, min(TOTAL_MATCHDAYS, matchday))

            matches_data = _fetch_json(
                f'https://api.football-data.org/v4/competitions/PL/matches?matchday={matchday}',
                headers,
            )
            standings_data = _fetch_json(
                'https://api.football-data.org/v4/competitions/PL/standings',
                headers,
            )
        except urllib.error.HTTPError as e:
            if e.code == 429:
                self._send_json(429, {'error': 'The football data provider is rate-limiting us right now — please wait a minute and try again.'})
            else:
                self._send_json(e.code, {'error': f'football-data.org returned {e.code} (check your API key)'})
            return
        except Exception as e:
            self._send_json(502, {'error': f'Fetch failed: {e}'})
            return

        events = []
        for m in matches_data.get('matches', []):
            home, away = m.get('homeTeam', {}), m.get('awayTeam', {})
            score = (m.get('score') or {}).get('fullTime') or {}
            state, desc = STATUS_MAP.get(m.get('status'), ('post', m.get('status', '')))
            events.append({
                'competitions': [{
                    'competitors': [
                        {
                            'team': {'id': home.get('id'), 'shortDisplayName': home.get('shortName') or home.get('name'), 'logo': home.get('crest')},
                            'score': score.get('home') if score.get('home') is not None else '-',
                        },
                        {
                            'team': {'id': away.get('id'), 'shortDisplayName': away.get('shortName') or away.get('name'), 'logo': away.get('crest')},
                            'score': score.get('away') if score.get('away') is not None else '-',
                        },
                    ],
                    'status': {'type': {'state': state, 'description': desc}},
                }],
            })

        table = ((standings_data.get('standings') or [{}])[0]).get('table', [])
        entries = []
        for row in table:
            team = row.get('team', {})
            entries.append({
                'team': {'id': team.get('id'), 'shortDisplayName': team.get('shortName') or team.get('name'), 'logos': [{'href': team.get('crest')}]},
                'stats': [
                    {'name': 'gamesPlayed', 'value': row.get('playedGames')},
                    {'name': 'wins', 'value': row.get('won')},
                    {'name': 'ties', 'value': row.get('draw')},
                    {'name': 'losses', 'value': row.get('lost')},
                    {'name': 'pointDifferential', 'value': row.get('goalDifference')},
                    {'name': 'points', 'value': row.get('points')},
                ],
            })

        self._send_json(200, {
            'scoreboard': {'events': events, 'matchday': matchday, 'totalMatchdays': TOTAL_MATCHDAYS},
            'standings': {'standings': [{'entries': entries}]},
        }, cache=True)

    def _send_json(self, status, payload, cache=False):
        body = json.dumps(payload).encode('utf-8')
        self.send_response(status)
        self.send_header('Content-Type', 'application/json')
        if cache:
            self.send_header('Cache-Control', 's-maxage=120, stale-while-revalidate=600')
        self.send_header('Content-Length', str(len(body)))
        self.end_headers()
        self.wfile.write(body)
